### PR TITLE
Fix a number of typescript errors

### PR DESCRIPTION
**Pull request by silverwind**
Fixes 96 typescript errors. Behaviour changes are commented below.

### CODE REVIEW

File: web_src/js/features/repo-code.ts [hunk 1]
```diff
@@ -142,13 +142,7 @@ export function initRepoCodeView() {
...
       }
       selectRange($(linesEls), $(selectedEls), from ? $(from) : null);
-
-      if (window.getSelection) {
-        window.getSelection().removeAllRanges();
-      } else {
-        document.selection.empty();
-      }
-
+      window.getSelection().removeAllRanges();
```
Comment (silverwind) [inline web_src/js/features/repo-code.ts:145]: I can't find anything on MDN regarding `document.selection`, so I think it never existed.

File: web_src/js/features/repo-graph.ts [hunk 2]
```diff
@@ -22,7 +22,7 @@ export function initRepoGraphGit() {
     for (const link of document.querySelectorAll('.pagination a')) {
       const href = link.getAttribute('href');
       if (!href) continue;
-      const url = new URL(href, window.location);
+      const url = new URL(href, window.location.href);
```
Comment (silverwind) [inline web_src/js/features/repo-graph.ts:25]: Typescript rejects passing Location to URL. `window.location.href` works the same.

File: web_src/js/features/repo-migration.ts [hunk 3]
```diff
@@ -25,11 +25,11 @@ export function initRepoMigration() {
...
   lfs?.addEventListener('change', setLFSSettingsVisibility);
 
-  const cloneAddr = document.querySelector('#clone_addr');
+  const cloneAddr = document.querySelector<HTMLInputElement>('#clone_addr');
   cloneAddr?.addEventListener('change', () => {
-    const repoName = document.querySelector('#repo_name');
+    const repoName = document.querySelector<HTMLInputElement>('#repo_name');
     if (cloneAddr.value && !repoName?.value) { // Only modify if repo_name input is blank
-      repoName.value = cloneAddr.value.match(/^(.*\/)?((.+?)(\.git)?)$/)[3];
+      repoName.value = /^(.*\/)?((.+?)(\.git)?)$/.exec(cloneAddr.value)[3];
```
Comment (silverwind) [inline web_src/js/features/repo-migration.ts:32]: Equivalent because regexp is not global.

File: web_src/js/features/repo-settings.ts [hunk 4]
```diff
@@ -122,7 +122,7 @@ function initRepoSettingsBranches() {
 function initRepoSettingsOptions() {
   if ($('.repository.settings.options').length > 0) {
     // Enable or select internal/external wiki system and issue tracker.
-    $('.enable-system').on('change', function () {
+    $('.enable-system').on('change', function (this: HTMLInputElement) {
```
Comment (silverwind) [inline web_src/js/features/repo-settings.ts:125]: FYI this is a [fake parameter](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-0.html#specifying-the-type-of-this-for-functions).

**Comment by wxiaoguang** [comment]
Please merge  Refactor issue filter (labels, poster, assignee) #32771  first, otherwise there will be a lot of conflicts.

**Comment by silverwind** [comment]
Merge was easy, only one small conflict because of the removed function.

**Review by wxiaoguang — APPROVED** [review]

**Review by lunny — APPROVED** [review]

**Pull request merged**